homeassistant/components/mqtt_room/sensor.py

```diff
@@ -118,48 +139,59 @@ def message_received(msg):
 
             device = _parse_update_data(msg.topic, data)
             if device.get(CONF_DEVICE_ID) == self._device_id:
-                if self._distance is None or self._updated is None:
-                    update_state(**device)
-                else:
+                for rec in self._room_states:
+                    if rec["distance"] is None or rec["updated"] is None:
+                        update_state(rec, **device)
+                        break
+
                     # update if:
                     # device is in the same room OR
                     # device is closer to another room OR
                     # last update from other room was too long ago
-                    timediff = dt.utcnow() - self._updated
+                    timediff = dt.utcnow() - rec["updated"]
                     if (
-                        device.get(ATTR_ROOM) == self._state
-                        or device.get(ATTR_DISTANCE) < self._distance
+                        device.get(ATTR_ROOM) == rec["room"]
+                        or device.get(ATTR_DISTANCE) < rec["distance"]
                         or timediff.total_seconds() >= self._timeout
                     ):
-                        update_state(**device)
+                        update_state(rec, **device)
+                        break
 
-        return await mqtt.async_subscribe(
-            self.hass, self._state_topic, message_received, 1
-        )
+        await mqtt.async_subscribe(self.hass, self._state_topic, message_received, 1)
 
     @property
-    def name(self):
+    def name(self) -> str:
         """Return the name of the sensor."""
         return self._name
 
     @property
     def extra_state_attributes(self):
         """Return the state attributes."""
-        return {ATTR_DISTANCE: self._distance}
+        primary_room_state = self._room_states[0]
+        attributes = {
+            ATTR_DISTANCE: primary_room_state["distance"],
+            ATTR_EXTRA_ROOMS: {},
+        }
+        for rec in self._room_states[1:]:
```

Comment:
What happens if there are more states reported then the number of extra_rooms?
Should we not omit the `extra_rooms` in the attributes if it is not set?
Please add some tests to make sure this is processed correctly.